File: study-nurio-mobile/ios/NotificationDestination.swift
import Foundation

enum NotificationDestination {
    private static let blockedPrefixes = [
        "/admin",
        "/tutoring",
        "/tutors",
    ]

    static func resolve(path: String?, url: String?, baseURL: URL) -> URL {
        guard let base = validBaseComponents(baseURL) else { return baseURL }
        let root = baseURL

        return resolveCandidate(path, base: base)
            ?? resolveCandidate(url, base: base)
            ?? root
    }

    private static func resolveCandidate(_ candidate: String?, base: URLComponents) -> URL? {
        guard let value = candidate?.trimmingCharacters(in: .whitespacesAndNewlines), !value.isEmpty,
              let components = URLComponents(string: value),
              components.fragment == nil,
              components.user == nil,
              components.password == nil else {
            return nil
        }

        let isRelativePath = components.scheme == nil && components.host == nil
        let isSameOriginURL = components.scheme?.lowercased() == "https" &&
            components.host?.lowercased() == base.host?.lowercased() &&
            effectivePort(components) == effectivePort(base)

        guard isRelativePath || isSameOriginURL else { return nil }

        let encodedPath = components.percentEncodedPath
        guard encodedPath.hasPrefix("/"),
              !encodedPath.hasPrefix("//"),
              !encodedPath.contains("%"),
              !isBlocked(encodedPath) else {
            return nil
        }

        return normalizedURL(path: encodedPath, query: components.percentEncodedQuery, base: base)
    }

    private static func validBaseComponents(_ baseURL: URL) -> URLComponents? {
        guard let components = URLComponents(url: baseURL, resolvingAgainstBaseURL: false),
              components.scheme?.lowercased() == "https",
              components.host?.isEmpty == false,
              components.user == nil,
              components.password == nil,
              components.fragment == nil,
              components.query == nil,
              components.path.isEmpty || components.path == "/",
              effectivePort(components) == 443 else {
            return nil
        }

        return components
    }

    private static func normalizedURL(path: String, query: String?, base: URLComponents) -> URL? {
        var normalized = URLComponents()
        normalized.scheme = "https"
        normalized.host = base.host?.lowercased()
        normalized.percentEncodedPath = path
        normalized.percentEncodedQuery = query
        return normalized.url
    }

    private static func effectivePort(_ components: URLComponents) -> Int {
        components.port ?? 443
    }

    private static func isBlocked(_ path: String) -> Bool {
        let normalized = path.lowercased()
        return blockedPrefixes.contains { prefix in
            normalized == prefix || normalized.hasPrefix("\(prefix)/")
        }
    }
}

File: study-nurio-mobile/android/app/src/main/java/com/nurio/study/android/notifications/NotificationDestination.kt
package com.nurio.study.android.notifications

import java.net.URI

object NotificationDestination {
    private val blockedPrefixes = listOf("/admin", "/tutoring", "/tutors")

    fun resolve(path: String?, url: String?, baseUrl: String): String {
        val base = parseBase(baseUrl) ?: return baseUrl
        val origin = origin(base)

        return resolveCandidate(path, base, origin)
            ?: resolveCandidate(url, base, origin)
            ?: origin
    }

    private fun resolveCandidate(candidate: String?, base: URI, origin: String): String? {
        val value = candidate?.trim()?.takeIf(String::isNotEmpty) ?: return null
        val uri = runCatching { URI(value) }.getOrNull() ?: return null

        if (uri.rawFragment != null || uri.rawUserInfo != null) return null

        val isRelativePath = uri.scheme == null && uri.rawAuthority == null
        val isSameOriginUrl = uri.scheme.equals("https", ignoreCase = true) &&
            uri.host?.equals(base.host, ignoreCase = true) == true &&
            effectivePort(uri) == effectivePort(base)

        if (!isRelativePath && !isSameOriginUrl) return null

        val rawPath = uri.rawPath.orEmpty()
        if (!rawPath.startsWith("/") || rawPath.startsWith("//") || rawPath.contains('%')) return null
        if (isBlocked(rawPath)) return null

        val query = uri.rawQuery?.let { "?$it" }.orEmpty()
        return "$origin$rawPath$query"
    }

    private fun parseBase(baseUrl: String): URI? {
        val base = runCatching { URI(baseUrl) }.getOrNull() ?: return null
        if (!base.scheme.equals("https", ignoreCase = true)) return null
        if (base.host.isNullOrBlank() || base.rawUserInfo != null || base.rawFragment != null) return null
        if (base.rawQuery != null || base.rawPath !in listOf("", "/")) return null
        if (effectivePort(base) != 443) return null
        return base
    }

    private fun origin(base: URI): String = "https://${base.host!!.lowercase()}"

    private fun effectivePort(uri: URI): Int = if (uri.port == -1) 443 else uri.port

    private fun isBlocked(path: String): Boolean {
        val normalized = path.lowercase()
        return blockedPrefixes.any { prefix ->
            normalized == prefix || normalized.startsWith("$prefix/")
        }
    }
}
